File: api-quarkus/src/test/java/br/ufrn/musi/EmMemoria.java
package br.ufrn.musi;

import br.ufrn.musi.aplicacao.DadosDaAnotacao;
import br.ufrn.musi.aplicacao.DadosDaObra;
import br.ufrn.musi.aplicacao.ErroDeAplicacao.Conflito;
import br.ufrn.musi.aplicacao.FiltroDeAnotacoes;
import br.ufrn.musi.aplicacao.FiltroDeObras;
import br.ufrn.musi.aplicacao.Pagina;
import br.ufrn.musi.aplicacao.PedidoDePagina;
import br.ufrn.musi.aplicacao.RepositorioDeAnotacoes;
import br.ufrn.musi.aplicacao.RepositorioDeObras;
import br.ufrn.musi.dominio.Dominio.Anotacao;
import br.ufrn.musi.dominio.Dominio.Obra;
import br.ufrn.musi.dominio.Dominio.Ordenacao;
import java.time.Instant;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

/**
 * Dublês (fakes) das portas, para testar os casos de uso sem banco: JUnit puro, sem
 * `@QuarkusTest`, em milissegundos. Mesmo papel de EmMemoria.kt do lado Ktor.
 */
final class EmMemoria {
    private EmMemoria() {}

    static final class Obras implements RepositorioDeObras {
        private final Map<String, Obra> obras = new LinkedHashMap<>();
        private int proximo = 1;

        Obras(Obra... iniciais) {
            for (Obra o : iniciais) obras.put(o.id(), o);
        }

        @Override public Pagina<Obra> listar(FiltroDeObras filtro, Ordenacao ordenacao, PedidoDePagina pedido) {
            var todas = new ArrayList<>(obras.values());
            int de = Math.min(pedido.pagina() * pedido.tamanho(), todas.size());
            return new Pagina<>(todas.subList(de, Math.min(de + pedido.tamanho(), todas.size())), pedido, todas.size());
        }
        @Override public Optional<Obra> buscar(String id) { return Optional.ofNullable(obras.get(id)); }
        @Override public boolean existe(String id) { return obras.containsKey(id); }
        @Override public Obra criar(DadosDaObra d) {
            if (d.mbid() != null && obras.values().stream().anyMatch(o -> d.mbid().equals(o.mbid())))
                throw new Conflito("já existe obra com este mbid");
            var o = obra("mem-" + proximo++, d);
            obras.put(o.id(), o);
            return o;
        }
        @Override public Optional<Obra> atualizar(String id, DadosDaObra d) {
            if (!obras.containsKey(id)) return Optional.empty();
            obras.put(id, obra(id, d));
            return Optional.of(obras.get(id));
        }
        @Override public boolean remover(String id) { return obras.remove(id) != null; }

        private static Obra obra(String id, DadosDaObra d) {
            return new Obra(id, d.titulo(), d.artista(), d.ano(), d.facetas(), d.mbid(), d.mbidComposicao());
        }
    }

    static final class Anotacoes implements RepositorioDeAnotacoes {
        private final List<Anotacao> anotacoes = new ArrayList<>();
        private long proximo = 1;

        @Override public Pagina<Anotacao> listar(String obraId, FiltroDeAnotacoes f, PedidoDePagina p) {
            var daObra = anotacoes.stream().filter(a -> a.obraId().equals(obraId)).toList();
            return new Pagina<>(daObra, p, daObra.size());
        }
        @Override public Optional<Anotacao> buscar(String obraId, long id) {
            return anotacoes.stream().filter(a -> a.id() == id && a.obraId().equals(obraId)).findFirst();
        }
        @Override public Anotacao criar(String obraId, DadosDaAnotacao d) {
            if (anotacoes.stream().anyMatch(a -> a.obraId().equals(obraId) && a.curador().equals(d.curador())
                    && Objects.equals(a.faceta(), d.faceta())))
                throw new Conflito("este curador já anotou esta faceta nesta obra");
            var a = new Anotacao(proximo++, obraId, d.faceta(), d.curador(), Instant.now());
            anotacoes.add(a);
            return a;
        }
        @Override public Optional<Anotacao> atualizar(String obraId, long id, DadosDaAnotacao d) {
            return buscar(obraId, id).map(a -> {
                var nova = new Anotacao(id, obraId, d.faceta(), d.curador(), a.criadoEm());
                anotacoes.set(anotacoes.indexOf(a), nova);
                return nova;
            });
        }
        @Override public boolean remover(String obraId, long id) {
            return anotacoes.removeIf(a -> a.id() == id && a.obraId().equals(obraId));
        }
    }
}

File: api-ktor/src/test/kotlin/br/ufrn/musi/EmMemoria.kt
package br.ufrn.musi

import br.ufrn.musi.aplicacao.BuscarObras
import br.ufrn.musi.aplicacao.Conflito
import br.ufrn.musi.aplicacao.DadosDaAnotacao
import br.ufrn.musi.aplicacao.DadosDaObra
import br.ufrn.musi.aplicacao.FiltroDeAnotacoes
import br.ufrn.musi.aplicacao.FiltroDeObras
import br.ufrn.musi.aplicacao.FonteDeObras
import br.ufrn.musi.aplicacao.Pagina
import br.ufrn.musi.aplicacao.PedidoDePagina
import br.ufrn.musi.aplicacao.RepositorioDeAnotacoes
import br.ufrn.musi.aplicacao.RepositorioDeObras
import br.ufrn.musi.dominio.Anotacao
import br.ufrn.musi.dominio.Faceta
import br.ufrn.musi.dominio.Filtro
import br.ufrn.musi.dominio.Obra
import br.ufrn.musi.dominio.Ordenacao
import br.ufrn.musi.dominio.buscar
import br.ufrn.musi.dominio.satisfaz
import org.koin.dsl.module
import kotlin.time.Clock

/**
 * Dublês (fakes) das portas, para testar casos de uso e rotas sem banco nem rede:
 * rodam em milissegundos e sem Docker. O comportamento com banco de verdade é conferido
 * em IntegracaoPostgresTest.
 */
class ObrasEmMemoria(vararg iniciais: Obra) : RepositorioDeObras {
    private val obras = iniciais.associateBy { it.id }.toMutableMap()
    private var proximo = 1

    override suspend fun listar(filtro: FiltroDeObras, ordenacao: Ordenacao, pedido: PedidoDePagina): Pagina<Obra> {
        val filtradas = obras.values.filter { o ->
            (filtro.artista == null || o.artista.contains(filtro.artista, ignoreCase = true)) &&
                (filtro.anoDe == null || o.ano >= filtro.anoDe) &&
                (filtro.anoAte == null || o.ano <= filtro.anoAte) &&
                (filtro.faceta == null || o.satisfaz(Filtro.Tem(filtro.faceta.dimensao, filtro.faceta.valor)))
        }.buscar(Filtro.E(emptyList()), ordenacao)
        return Pagina(filtradas.drop(pedido.deslocamento.toInt()).take(pedido.tamanho), pedido, filtradas.size.toLong())
    }

    override suspend fun buscar(id: String) = obras[id]
    override suspend fun existe(id: String) = id in obras

    override suspend fun criar(dados: DadosDaObra): Obra {
        exigirMbidLivre(dados.mbid, null)
        return dados.paraObra("mem-${proximo++}").also { obras[it.id] = it }
    }

    override suspend fun atualizar(id: String, dados: DadosDaObra): Obra? {
        if (id !in obras) return null
        exigirMbidLivre(dados.mbid, id)
        return dados.paraObra(id).also { obras[id] = it }
    }

    override suspend fun remover(id: String) = obras.remove(id) != null

    private fun exigirMbidLivre(mbid: String?, excetoId: String?) {
        if (mbid != null && obras.values.any { it.mbid == mbid && it.id != excetoId })
            throw Conflito("já existe obra com este mbid")
    }

    private fun DadosDaObra.paraObra(id: String) = Obra(id, titulo, artista, ano, facetas, mbid, mbidComposicao)
}

class AnotacoesEmMemoria : RepositorioDeAnotacoes {
    private val anotacoes = mutableListOf<Anotacao>()
    private var proximo = 1L

    override suspend fun listar(obraId: String, filtro: FiltroDeAnotacoes, pedido: PedidoDePagina): Pagina<Anotacao> {
        val filtradas = anotacoes.filter {
            it.obraId == obraId &&
                (filtro.curador == null || it.curador == filtro.curador) &&
                (filtro.dimensao == null || it.faceta.dimensao == filtro.dimensao)
        }
        return Pagina(filtradas.drop(pedido.deslocamento.toInt()).take(pedido.tamanho), pedido, filtradas.size.toLong())
    }

    override suspend fun buscar(obraId: String, id: Long) = anotacoes.find { it.id == id && it.obraId == obraId }

    override suspend fun criar(obraId: String, dados: DadosDaAnotacao): Anotacao {
        exigirUnica(obraId, dados, null)
        return Anotacao(proximo++, obraId, dados.faceta, dados.curador, Clock.System.now()).also { anotacoes += it }
    }

    override suspend fun atualizar(obraId: String, id: Long, dados: DadosDaAnotacao): Anotacao? {
        val atual = buscar(obraId, id) ?: return null
        exigirUnica(obraId, dados, id)
        return atual.copy(faceta = dados.faceta, curador = dados.curador).also { anotacoes[anotacoes.indexOf(atual)] = it }
    }

    override suspend fun remover(obraId: String, id: Long) = anotacoes.removeIf { it.id == id && it.obraId == obraId }

    private fun exigirUnica(obraId: String, d: DadosDaAnotacao, excetoId: Long?) {
        if (anotacoes.any { it.obraId == obraId && it.curador == d.curador && it.faceta == d.faceta && it.id != excetoId })
            throw Conflito("este curador já anotou esta faceta nesta obra")
    }
}

/** O serviço Go substituído por uma lista: a busca avalia a árvore com o domínio. */
class FonteEmMemoria(private val acervo: List<Obra>) : FonteDeObras {
    override suspend fun buscar(filtro: Filtro) = acervo.filter { it.satisfaz(filtro) }
}

val ASA_BRANCA = Obra("obra-03", "Asa Branca", "Luiz Gonzaga", 1947,
    listOf(Faceta("genero", "forro"), Faceta("ritmo", "baiao")))
val PONTEIO = Obra("obra-01", "Ponteio", "Edu Lobo", 1967,
    listOf(Faceta("genero", "mpb"), Faceta("ritmo", "ponteio")), mbid = "11111111-1111-4111-8111-111111111111")

/** Os módulos de Koin com as portas em memória, no lugar do Go e do PostgreSQL. */
fun modulosEmMemoria(vararg obras: Obra) = listOf(
    module {
        single<FonteDeObras> { FonteEmMemoria(obras.toList()) }
        single { BuscarObras(get()) }
        single<RepositorioDeObras> { ObrasEmMemoria(*obras) }
        single<RepositorioDeAnotacoes> { AnotacoesEmMemoria() }
    },
    casosDeUso,
)
